### ISSUE TITLE
Enable plotting of timedeltas

### ISSUE THREAD

**Issue by dstansby** [matplotlib/matplotlib#8869]
This is a general issue to track progress in enabling the plotting of `timedelta` objects. The eventual aim is for code like this to work:

```python
import matplotlib.pyplot as plt
from datetime import timedelta

x = [1, 2]
y = [timedelta(seconds=1), timedelta(seconds=2)]
fig, ax = plt.subplots()
ax.scatter(x, y)
plt.show()
```

TODO
------

- [ ] Method to convert `timedelta` to Matplotlib's internal time/date representation (float number of days) - #8730
- [x] Method to convert Matplotlib's internal time/date representation back to `timedelta` - #8870
- [ ] Implement a `TimedeltaFormatter` - #8930
- [ ] Implement a `TimedeltaLocator` (possibly along with `timedelta` versions of `SecondLocator`, `MinuteLocator` etc.)
- [ ] Implement an `AutoTimedeltaLocator`
- [ ] Implement a `TimedeltaConverter` to go in the units registry

**Comment by anntzer** [comment]
Quick question from the peanut gallery: could datetime/timedelta support be folded into unit support?  (I have no idea.)

**Comment by dstansby** [comment]
(disclaimer: I don't really know what I'm doing with all this and am just copying `datetime` support and learning as I go along)

I think `datetime` is registered in the unit framework. The last line of `dates.py` is `units.registry[datetime.datetime] = DateConverter()` which means a datetime passed to plot automatically uses all the converter/ticker stuff in `dates.py`. Is that what you meant?

**Comment by anntzer** [comment]
I don't know anything about datetime plotting either, but the general idea was indeed whether we could end up with something like units.registry[timedelta] = TimeDeltaConverter.

**Comment by has2k1** [comment]
I dealt with timedelta [breaks](https://mizani.readthedocs.io/en/stable/breaks.html#mizani.breaks.timedelta_breaks) and [formats](
https://mizani.readthedocs.io/en/stable/formatters.html#mizani.formatters.timedelta_format) in [mizani](https://github.com/has2k1/mizani). The implementation is not the cleanest but I have been satisfied with the results.

**Comment by pganssle** [comment]
Is there something somewhere that details how unit support is handled? Folding `timedelta` directly into `datetime` seems weird - it seems like it would enable you to do things like `plt.plot([timedelta(days=200), datetime(1970, 11, 14)], [0, 1])`? Seems like you'd want to be kinda explicit about when you can use `timedelta` and when you can use `datetime`.

That said, we're essentially *never* explicit about when you can use "meters" vs. "delta-meters", so maybe my concerns are a bit overblown? I guess it feels weirder because numbers are mostly unitless and these have explicit, non-interchangeable units.

**Comment by dstansby** [comment]
Thanks for all your comments @pganssle!

I think there are two cases that I've probably accidentally folded into one here:

1. Plotting `timedelta` objects
2. Plotting `datetime` objects that need some sort of offsetting with `timedelta` (ie. plotting a square

In case number 2., offsetting should be done before any conversion to internal representations, such that only `datetime` coordinates are ever plotted, and none of the converting `timedelta` machinery is needed.

Case 1. is what should be tackled by this issue, and as is said above, we need to be very careful to keep the two cases distinct, and error when a code snippet like above (https://github.com/matplotlib/matplotlib/issues/8869#issuecomment-323498027) is run.

**Comment by pganssle** [comment]
@dstansby Yes, I think with these concerns cleanly separated, implementing `timedelta` handling becomes much easier, too. Other than the question of formatting (and if you have any references on timedelta formatting [I'd like to implement that in dateutil](https://github.com/dateutil/dateutil/issues/444), so I'm interested in what's the "best way"), so long as the units are being enforced, it's just `timedelta.total_seconds()`, none of the complicated nonsense about time zones and the like, since `timedelta()` by its nature represents a duration.

**Comment by dstansby** [comment]
Great, will put a comment on that issue about timedelta formatting. Would be much better to have it somewhere in `dateutil` than for us to roll our own version.

**Comment by dstansby** [comment]
Was easier to just open a new PR at #9120

**Comment by dstansby** [comment]
Woops, that comment was meant for the PR!

**Comment by jklymak** [comment]
We've tried this a few times - but after thinking about it for a few years I'm not sure we want or need this.  

timedelta are just integers or floats with a unit attached.  
```
dt = np.array([1, 20, 30], dtype='timedelta64[s]')
ax.plot(dt)
```
works fine as-is.  

For `datetime.timedelta` we would need to convert to float:

```
dt0 = [datetime.timedelta(seconds=i) for i in [1, 20, 30]]
dt = [d.total_seconds() for d in dt0]
ax.plot(dt)
```

Personally I'm OK with `datetime.timedelta` being a second-class citizen.  If we wanted to have a _really_ light converter for it that just converted the values to float for the user, but otherwise didn't change the Locators and Formatters, that would probably be OK too.  

The ideas in #19236 were that we would use datetime-like Formatters and Locators to the timedeltas, but I'm no longer convinced that this is a good idea.  If someone really wants time, they can add an origin datetime to each timedelta, either when their experiment started, or a nominal date and pass to our datetime converter.

**Comment by gitsdfawfevi** [comment]
> The ideas in #19236 were that we would use datetime-like Formatters and Locators to the timedeltas, but I'm no longer convinced that this is a good idea. If someone really wants time, they can add an origin datetime to each timedelta, either when their experiment started, or a nominal date and pass to our datetime converter.

I evaluate simulations and write the values into a pandas DataFrame with timedelta as index. Using datetime is the wrong choice in this case, as there is no starting day on which the simulation starts (it is relatively at 0 seconds).  see my whole comment: https://github.com/pandas-dev/pandas/issues/60079#issuecomment-2467898781

**Comment by jklymak** [comment]
@gitsdfawfevi Yes, we understand that use case.  The question is if there is truly value in HH.MM.SS since simulation start, or if folks are OK with decimal seconds, hours, or days etc. Most fields I'm familiar with, the latter is fine.

**Comment by gitsdfawfevi** [comment]
@jklymak I assume that you mean the following representation 03:15:00 == 3.25 h, therefore an example of when the decimal representation is very unfavorable in a plot: For example, there is data from a year simulation, represented decimal in months (= 12 mo). Now you want to interactively enlarge the plot to a certain time range in seconds. Of course, this can also be achieved by converting units and changing the corresponding axis labeling, but it is very complex.

Regardless of whether this is implemented or not, I find it very critical how matplotlib deals with timedelta “The critical point is that matplotlib pretends to understand timedelta, but the result is ‘wrong’ -- at least not what you would expect.” (https://github.com/pandas-dev/pandas/issues/60079#issuecomment-2476601817)

**Comment by jklymak** [comment]
lets not mix concerns here - Matplotlib deals with `np.timedelta64`, we don't directly handle pandas Timedelta plotting, and we don't provide `pd.df.plot`.  If there is something we can do to make that work more smoothly, we will, because pandas is used so extensively, but please express the feature request in terms of numpy objects and our plotting routines (which underpin the pandas ones)

**Comment by gitsdfawfevi** [comment]
Sorry, that was my mistake. I wasn't aware that there is a difference between pandas and numpy timedelta. Especially as both use the same dtype designation.

In example https://github.com/matplotlib/matplotlib/issues/8869?notification_referrer_id=MDE4Ok5vdGlmaWNhdGlvblRocmVhZDEzMzYzNzkyNTkwOjEzOTk1MzQ1MQ%3D%3D#issue-242341490 , timedelta from python datetime is used, but it is also different from numpy datetime (https://numpy.org/doc/stable/reference/arrays.datetime.html). Would the implementations of datetime from numpy and python then be compatible with matplotlib?

**Comment by maoding** [comment]
> lets not mix concerns here - Matplotlib deals with `np.timedelta64`, we don't directly handle pandas Timedelta plotting, and we don't provide `pd.df.plot`. If there is something we can do to make that work more smoothly, we will, because pandas is used so extensively, but please express the feature request in terms of numpy objects and our plotting routines (which underpin the pandas ones)

@jklymak the example with numpy objects is the following:

```python
import numpy as np
import matplotlib.pyplot as plt

num_values = 10
now = np.datetime64("2005-02-25")

datetime_values = np.array(
    [now + np.timedelta64(i, "h") * i for i in range(num_values)]
)
timedelta_values = np.diff(datetime_values)
zero_timedelta = np.array([np.timedelta64(0, "h")])
timedelta_values = np.cumsum(np.insert(timedelta_values, 0, zero_timedelta))
values = [i for i in range(num_values)]

plt.figure()
plt.plot(datetime_values, values, marker=".")
plt.xticks(rotation=90)
plt.title("Numpy Datetime")
plt.tight_layout()
plt.show()
plt.close(plt.gcf())

plt.figure()
plt.plot(timedelta_values, values, marker=".")
plt.xticks(rotation=90)
plt.title("Numpy Timedelta")
plt.tight_layout()
plt.show()
plt.close(plt.gcf())
```

Resulting in 
![image](https://github.com/user-attachments/assets/c916a916-2f1e-4105-a5c7-190b4d6d0ddf)
and
![image](https://github.com/user-attachments/assets/0478fc84-3617-41e1-aa05-effd5794da5c)

Here the context is displayed correctly in both cases. In https://github.com/pandas-dev/pandas/issues/60079 the context isn't displayed properly, compare the linear context with `pd.timedelta` versus the curved with `pd.datetime`.

So this is a problem of pandas using matplotlib (or anythin related) rather than a matplotlib problem, and thus should be fixed in pandas, isn't it?

**Comment by scottshambaugh** [comment]
> 2) Plotting datetime objects that need some sort of offsetting with timedelta

I'm running into this use case right now. Allowing `Axes.convert_xunits` and `Axes.convert_yunits` to take in `timedelta` objects seems reasonable, for axes that already have a time scale.

**Comment by scottshambaugh** [comment]
As a workaround, the below shows how to convert a timedelta to matplotlib's numerical date coordinates.

```python
import matplotlib.dates as mdates
from matplotlib.lines import Line2D
from datetime import datetime, timedelta

line: Line2D
x_date: datetime
x_offset: timedelta

x_num = line.convert_xunits(x_date)
xoffset = mdates.date2num(x_offset + x_date) - x_num
```

**Comment by jklymak** [comment]
@scottshambaugh Thanks for this.  Can you clarify the output you want and how you originally set the x-axis units?  

Thanks,   Jody

**Comment by scottshambaugh** [comment]
Hm, that was driven by some subclassing of `Text` and manually setting positions. When working through the normal interfaces, datetimes as coordinates work fine. I'm having some trouble making a standalone script that shows the issue I was running into, it's probably fine as-is.
